### PR TITLE
SamplingPriority refactor and support for UserKeep and UserDrop

### PR DESCRIPTION
**Pull request by willgittoes-dd**
The feature added is UserKeep and UserDrop (although they cannot yet easily be set by a user without compiling against our headers). The primary delta/change for this PR however, is a refactoring of how SamplingPriority is handled (set, assigned, and propagated).

Previously, SamplingPriority was a property of the SpanContext - which is not logical since it applies to a trace as a whole not a single span. However, it was there because it propagates via SpanContext. Now, SamplingPriority is handled at the Trace level, which means it is handled by the SpanBuffer, which contains the PendingTrace. When a PendingTrace completes, then the SamplingPriority is assigned (if relevant). 

This should allow better behaviour of when users may apply UserKeep and UserDrop: at any time before a trace propagates or the root span finishes.

Also:

* Renamed a random method from snake_case to camelCase.
* Added more verbosity to tsan output.

### CODE REVIEW

**Comment by willgittoes-dd** [comment]
Re: sampling.priority, YES! Since that's what rnburn asked for.

Re: the second question, not sure what you mean.

**Comment by cgilmour** [comment]
> Re: the second question, not sure what you mean.
Never mind, I'd misunderstood the lifecycle of things in the pending traces map.

**Comment by willgittoes-dd** [comment]
Sampling priority tag added, please review the latest commit!

File: src/span.cpp [hunk 1]
```diff
@@ -18,6 +19,7 @@ const std::string datadog_resource_name_tag = "resource.name";
 const std::string datadog_service_name_tag = "service.name";
 const std::string http_url_tag = "http.url";
 const std::string operation_name_tag = "operation";
+const std::string priority_sampling_tag = "sampling.priority";
```
Comment (cgilmour) [inline src/span.cpp]: The tag is defined by opentracing in `opentracing/ext/tags.h`.

File: test/propagation_test.cpp [hunk 2]
```diff
@@ -167,13 +167,31 @@ TEST_CASE("sampling behaviour") {
   std::shared_ptr<Tracer> tracer{new Tracer{tracer_options, buffer, getRealTime, getId, sampler}};
   ot::Tracer::InitGlobal(tracer);
 
+  // There's two ways we can set the sampling priority. Either directly using the method, or
+  // through a tag. Test both.
+  auto setSamplingPriority = GENERATE(
```
Comment (cgilmour) [inline test/propagation_test.cpp]: Fancy!

File: test/propagation_test.cpp [hunk 3]
```diff
@@ -167,13 +167,31 @@ TEST_CASE("sampling behaviour") {
...
+      values<
+          std::function<OptionalSamplingPriority(Span*, std::unique_ptr<UserSamplingPriority>)>>({
+          [](Span* span, std::unique_ptr<UserSamplingPriority> p) {
+            return span->setSamplingPriority(std::move(p));
+          },
+          [](Span* span, std::unique_ptr<UserSamplingPriority> p) {
+            if (p != nullptr) {
+              span->SetTag("sampling.priority", static_cast<int>(*p));
+            } else {
+              span->SetTag("sampling.priority", "");
```
Comment (cgilmour) [inline test/propagation_test.cpp]: Should it just be unset?

**Comment by cgilmour** [comment]
A few review comments.

File: src/span.cpp (hunk 1 above)
Comment (willgittoes-dd) [reply src/span.cpp]: Tru dat.

File: test/propagation_test.cpp (hunk 2 above)
Comment (willgittoes-dd) [reply test/propagation_test.cpp]: yeah innit?!

File: test/propagation_test.cpp (hunk 3 above)
Comment (willgittoes-dd) [reply test/propagation_test.cpp]: That's what this does. So in addition to the spec of "0 = discard, anything else = keep" I'm adding a way to unset the sampling priority, by passing in an empty string.

**Pull request merged**